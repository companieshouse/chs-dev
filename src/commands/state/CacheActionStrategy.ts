import yaml from "yaml";

import { getGeneratedDockerComposeFile } from "../../helpers/docker-compose-file.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import { writeContentToFile } from "../../helpers/file-utils.js";
import { handlePrompt, validateCacheNameExists } from "./state-cache-utils.js";
import { StateManager } from "../../state/state-manager.js";

const EXPORT_STATE_DIR = ".exported_state_cache";

export interface AvailableCacheContext {
    logger: (message: string) => void;
  }

export interface WipeCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface RemoveCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface ExportCacheContext {
    projectPath: string;
    logger: (message: string) => void;
}

export interface AddCacheContext {
    stateManager: StateManager;
    stateCacheFile: string;
    projectPath: string;
    logger: (message: string) => void;
}
interface CacheActionStrategy<C> {
    execute(cache: C, cacheData: Record<string, any>, cacheName?: string): Promise<void> | void;
}

const handleExportCache = (context: ExportCacheContext, cacheData: Record<string, any>, cacheName: string): void => {
    const exportDir = join(context.projectPath, EXPORT_STATE_DIR);
    if (!existsSync(exportDir)) {
        mkdirSync(exportDir, { recursive: true });
    }

    const exportedFilenamePath = join(exportDir, `${cacheName}.yaml`);

    writeContentToFile(cacheData[cacheName], exportedFilenamePath);
    context.logger(`Exported cache ${cacheName} destination: '${exportedFilenamePath}'`);
};

export class AvailableCacheStrategy implements CacheActionStrategy<AvailableCacheContext> {
    execute (context: AvailableCacheContext, cacheData: Record<string, any>): void {
        const cacheNames = Object.keys(cacheData);
        if (cacheNames.length > 0) {
            context.logger("Available caches:");
            cacheNames.forEach((name: string) => context.logger(`- ${name}`));
        } else {
            context.logger("No cache available.");
        }
    }
}

export class WipeCacheStrategy implements CacheActionStrategy<WipeCacheContext> {
    async execute (context: WipeCacheContext, _cacheData: Record<string, any>): Promise<void> {
        const confirmed = await handlePrompt("wipe");
        if (confirmed) {
            writeContentToFile({}, context.stateCacheFile);
            context.logger("Wiped all caches");
        }
    }
}

export class RemoveCacheStrategy implements CacheActionStrategy<RemoveCacheContext> {
    async execute (context: RemoveCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);

        const confirmed = await handlePrompt("remove", _cacheName);
        if (confirmed) {
            delete _cacheData[_cacheName];
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Removed cache '${_cacheName}'`);

        }
    }
}

export class ExportCacheStrategy implements CacheActionStrategy<ExportCacheContext> {
    async execute (context: ExportCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);
        handleExportCache(context, _cacheData, _cacheName);

    }
}

export class AddCacheStrategy implements CacheActionStrategy<AddCacheContext> {
    async execute (context: AddCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        const confirmed = await handlePrompt("add", _cacheName);
        if (confirmed) {
            const stateSnapshot = context.stateManager.snapshot;
            const dockerComposeSnapshot = getGeneratedDockerComposeFile(context.projectPath);
            _cacheData[_cacheName] = {
                state: {
                    hash: hashAlgorithm(yaml.stringify(stateSnapshot)),
                    snapshot: stateSnapshot
                },
                dockerCompose: {
                    hash: hashAlgorithm(yaml.stringify(dockerComposeSnapshot)),
                    snapshot: dockerComposeSnapshot
                }
            };
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Saved cache as '${_cacheName}'`);
        }
    }
}
