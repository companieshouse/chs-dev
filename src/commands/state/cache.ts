import { Args, Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Cache extends Command {

    static description = "Cache the state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { args, flags } = await this.parse(Cache);
        const timestamp: string = new Date(Date.now()).toISOString();
        const cacheName: string = args.name || timestamp;

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`save chs-dev state?`);
        };

        if (await handlePrompt()) {
            //
        }
    }

}
