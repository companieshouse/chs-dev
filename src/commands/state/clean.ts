import { Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Clean extends Command {

    static description = "Wipes the state of chs-dev and revert to a default state";

    static flags = {
        purge: Flags.boolean({
            char: "p",
            description: "Wipe state, volumes and images",
            default: false
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { flags: { purge } } = await this.parse(Clean);

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`This will reset your chs-dev state and delete all unused docker volumes${purge ? " including images." : "."} Do you want to proceed?`);
        };

        if (await handlePrompt()) {
            const containerStatus = this.dockerCompose.getServiceStatuses();
            if (containerStatus !== undefined) {
                this.error("Ensure all containers are stopped. Run: '$ chs-dev down'");
            }
            this.stateManager.cleanState();

            await this.config.runHook("generate-runnable-docker-compose", {});

            this.dockerCompose.prune("volume");
            if (purge) {
                this.dockerCompose.prune("image");
            }

        }
    }

}
