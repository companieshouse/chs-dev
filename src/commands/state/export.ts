import { Args, Command, Config } from "@oclif/core";
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import yaml from "yaml";
import loadConfig from "../../helpers/config-loader.js";
import { getGeneratedDockerComposeFile } from "../../helpers/docker-compose-file.js";
import { writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { confirm } from "../../helpers/user-input.js";
import ChsDevConfig from "../../model/Config.js";
import { StateManager } from "../../state/state-manager.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string
    }
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    }
};
const EXPORT_STATE_DIR = ".exported_state_cache";

export default class Export extends Command {
    static description = "Export current state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: true,
            description: "Name of the exported file"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly exportedStateDir: string;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        this.chsDevConfig = loadConfig();
        this.exportedStateDir = join(this.chsDevConfig.projectPath, EXPORT_STATE_DIR);

        if (!existsSync(this.exportedStateDir)) {
            mkdirSync(this.exportedStateDir, { recursive: true });
        }

        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
    }

    async run (): Promise<void> {
        const { args: { name } } = await this.parse(Export);

        if (await confirm(`Export current state of chs-dev to file named: '${name}'?`)) {
            this.exportState(name);
        }

    }

    private exportState (exportFileName: string): void {
        const exportedFilenamePath = join(this.exportedStateDir, `${exportFileName}.yaml`);
        writeContentToFile(this.getExportData(), exportedFilenamePath);

        this.log(`Exported state destination: ${exportedFilenamePath}`);
    }

    private getExportData (): StateCache {
        const stateSnapshot = this.stateManager.snapshot;
        const dockerComposeSnapshot = getGeneratedDockerComposeFile(this.chsDevConfig.projectPath);
        return {
            state: {
                hash: hashAlgorithm(yaml.stringify(stateSnapshot)),
                snapshot: stateSnapshot
            },
            dockerCompose: {
                hash: hashAlgorithm(yaml.stringify(dockerComposeSnapshot)),
                snapshot: dockerComposeSnapshot
            }
        };

    }

}
