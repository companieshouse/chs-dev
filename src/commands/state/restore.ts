import { Args, Command, Config, Flags } from "@oclif/core";
import { existsSync } from "fs";
import { basename, join } from "path";
import yaml from "yaml";
import loadConfig from "../../helpers/config-loader.js";
import { readFileContent, writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { confirm } from "../../helpers/user-input.js";
import ChsDevConfig from "../../model/Config.js";
import { DockerCompose } from "../../run/docker-compose.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string
    }
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    }
};

export default class Restore extends Command {
    static description = "Restore and update the state files from saved cache or imported cache.";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    static flags = {
        importCacheFrom: Flags.string({
            char: "i",
            description: "Path to the exported cache"
        })
    };

    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;
    private readonly stateCacheFile: string;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        this.chsDevConfig = loadConfig();
        const cacheDir = config.cacheDir;
        this.stateCacheFile = join(cacheDir, `${basename(this.chsDevConfig.projectName)}.state.yaml`);

        const logger = { log: (msg: string) => this.log(msg) };

        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);
    }

    async run (): Promise<void> {
        const { args, flags: { importCacheFrom } } = await this.parse(Restore);

        const containerStatus = this.dockerCompose.getServiceStatuses();
        if (containerStatus !== undefined) {
            this.error("Ensure all containers are stopped. Run: '$ chs-dev down'");
        }

        const cacheName = importCacheFrom ? "import-cache" : args.name;

        if (!cacheName || typeof cacheName === "undefined") {
            this.error("Please provide a valid cache name or use the --importCacheFrom or -i flag to restore from an imported cache.");
        }
        console.log(`Restoring state from cache: ${cacheName}`);
        if (importCacheFrom) {
            if (await this.handlePrompt(cacheName)) {
                this.restoreFromImport(importCacheFrom);
            }
        } else {
            if (await this.handlePrompt(cacheName)) {
                this.restoreFromSavedCache(cacheName);
            }
        }

    }

    private restoreFromSavedCache (cacheName: string): void {
        const cacheData = this.loadSavedCacheData();
        const namedCacheData = cacheData[cacheName];
        if (!namedCacheData) {
            this.error(`Cache with name '${cacheName}' not found.`);
        }
        const data = this.verifyCacheAuthencity(namedCacheData, (stateCache) => stateCache);

        this.restoreState(data);
        this.log(`Restored state from saved cache '${cacheName}'`);
    }

    private restoreFromImport (importCachePath: string): void {
        if (!existsSync(importCachePath)) {
            this.error(`Import cache file does not exist in location: ${importCachePath}`);
        }

        const cacheData = readFileContent(importCachePath) as StateCache;

        if (!cacheData || typeof cacheData !== "object") {
            this.error(`Invalid cache data in imported file: ${importCachePath}`);
        }

        const data = this.verifyCacheAuthencity(cacheData, (stateCache) => stateCache);

        this.restoreState(data);
        this.log(`Restored state from imported cache.`);
    }

    private verifyCacheAuthencity<T> (cacheData: StateCache, cacheSupplier: (stateCache: StateCache) => StateCache): StateCache {

        if (cacheData.state.hash === hashAlgorithm(yaml.stringify(cacheData.state.snapshot)) &&
            cacheData.dockerCompose.hash === hashAlgorithm(yaml.stringify(cacheData.dockerCompose.snapshot))) {
            return cacheSupplier(cacheData);
        }
        this.error("Cache data has been corrupted or touched.");
    }

    private restoreState ({ state, dockerCompose }: StateCache): void {
        // Restore state file
        const stateFilePath = join(this.chsDevConfig.projectPath, `.chs-dev.yaml`);
        writeContentToFile(state.snapshot, stateFilePath);

        // Restore docker-compose file
        const dockerComposeFilePath = join(this.chsDevConfig.projectPath, "docker-compose.yaml");
        writeContentToFile(dockerCompose.snapshot, dockerComposeFilePath);
    }

    private loadSavedCacheData (): Record<string, StateCache> {
        return readFileContent(this.stateCacheFile) as Record<string, StateCache> || {};
    }

    private async handlePrompt (cacheName: string): Promise<boolean> {
        const message = cacheName === "import-cache"
            ? `Restore from an imported cache?`
            : `Restore from a saved cache '${cacheName}'?`;
        return confirm(message);
    }
}
