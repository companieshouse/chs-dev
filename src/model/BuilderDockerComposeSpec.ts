/**
 * Represents a Docker Compose specification for a builder
 */
export interface BuilderDockerComposeSpec {
    /**
     * Name of the builder
     */
    name: string;

    /**
     * Version of the builder
     */
    version: string;

    /**
     * String representation of the docker compose spec - may contain variables
     * in '< >' which are removed when the docker compose file is created
     */
    builderSpec: string;
}

export default BuilderDockerComposeSpec;
