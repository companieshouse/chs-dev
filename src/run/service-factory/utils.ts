import NewServiceSpec from "../../model/NewServiceSpec.js";

/**
 * Apply the value to the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param selector Jq like selector to apply the value to
 * @param newServiceSpec The newServiceSpec to apply the value to
 * @param newValue The value to apply
 * @returns The newServiceSpec with the value applied
 */
export const applySelector = (newServiceSpec: Partial<NewServiceSpec>, selector: string, newValue: any) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return {
            ...newServiceSpec,
            [nestedKeys[0]]: applySelector(
                newServiceSpecSubAttribute,
                `.${nestedKeys.slice(1).join(".")}`,
                newValue
            )
        };
    } else {
        return {
            ...newServiceSpec,
            [nestedKeys[0]]: newValue
        };
    }
};

/**
 * Lookup a value in the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param newServiceSpec NewServiceSpec to lookup the value in
 * @param selector Jq like selector to lookup the value at
 * @returns The value at the selector
 */
export const get = (newServiceSpec: Partial<NewServiceSpec>, selector: string) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return get(newServiceSpecSubAttribute, `.${nestedKeys.slice(1).join(".")}`);
    } else {
        return newServiceSpec[nestedKeys[0]];
    }
};

const isSelectorValid = (selector: string) => {
    return /^\.([A-Za-z0-9_-]+\.?)+$/.test(selector);
};

const extractValidNestedKeys = (selector: string) => {
    if (!isSelectorValid(selector)) {
        throw new Error("Invalid selector");
    }

    return selector.split(".").slice(1);
};

const getNextLevelObject = (newServiceSpec: Partial<NewServiceSpec>, nestedKey: string) => {
    return newServiceSpec[nestedKey] || {};
};
