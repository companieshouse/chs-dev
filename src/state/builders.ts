import { Dirent, readdirSync, readFileSync } from "fs";
import { join } from "path";
import BuilderDockerComposeSpec from "../model/BuilderDockerComposeSpec.js";
import yaml from "yaml";

/**
 * Loads the requested version of the builder specification from the inventory
 * @param projectPath path to the project the tool is being run out of
 * @param builderName name of the builder being requested
 * @param version version of the builder being used
 * @returns BuilderDockerComposeSpec representing the builder
 */
export const getBuilder = (projectPath: string, builderName: string | undefined, version: string | undefined): BuilderDockerComposeSpec => {
    if (typeof builders === "undefined") {
        builders = {
            repository: {
                v1: {
                    name: "repository",
                    version: "v1",
                    builderSpec: repositoryBuilderSpec
                }
            }
        };

        populateBuildersCache(projectPath, builders);
    }

    if (typeof builderName === "undefined" || builderName.length === 0) {
        return getBuilder(projectPath, "repository", undefined);
    }

    if (!(builderName in builders)) {
        throw new Error(`No builder found for ${builderName} builder`);
    }

    const builderVersions = builders[builderName];

    const builderVersion = typeof version === "undefined"
        ? builderVersions[findLatestVersion(builderVersions)]
        : builderVersions[version];

    if (typeof builderVersion === "undefined") {
        throw new Error(`No version ${version} of ${builderName} builder`);
    }

    return builderVersion;
};

export const getBuilders = (projectPath: string) => {
    if (typeof builders === "undefined") {
        builders = {
            repository: {
                v1: {
                    name: "repository",
                    version: "v1",
                    builderSpec: repositoryBuilderSpec
                }
            }
        };

        populateBuildersCache(projectPath, builders);
    }

    return {
        ...builders
    };
};

export const clearBuilders = () => {
    builders = undefined;
};

const repositoryBuilderSpec = yaml.stringify({
    services: {
        "<service>": {
            pull_policy: "build",
            build: {
                context: "<absolute_repository_path>",
                args: {
                    // eslint-disable-next-line no-template-curly-in-string
                    SSH_PRIVATE_KEY: "${SSH_PRIVATE_KEY}",
                    // eslint-disable-next-line no-template-curly-in-string
                    SSH_PRIVATE_KEY_PASSPHRASE: "${SSH_PRIVATE_KEY_PASSPHRASE}"
                }
            },
            develop: {
                watch: [
                    {
                        path: ".touch",
                        action: "rebuild"
                    }
                ]
            }
        }
    }
});

let builders: undefined | Record<string, Record<string, BuilderDockerComposeSpec>>;

const isBuilderFile = (potentialBuilderFile: Dirent) => {
    return potentialBuilderFile.isFile() && (
        potentialBuilderFile.name === "builder.docker-compose.yaml" ||
        potentialBuilderFile.name === "builder.docker-compose.yml"
    );
};
const findLatestVersion = (builderVersions: Record<string, BuilderDockerComposeSpec>) => {
    return Object.keys(builderVersions).sort().reverse()[0];
};

const populateBuildersCache = (projectPath: string, builders: Record<string, Record<string, BuilderDockerComposeSpec>>) => {
    const buildersDir = join(projectPath, "local/builders");
    const buildersDirContents: Dirent[] = readdirSync(
        buildersDir, {
            recursive: true,
            withFileTypes: true
        }
    );

    const builderFiles = buildersDirContents.filter(isBuilderFile);

    for (const builder of builderFiles) {
        const pathWithoutBase = builder.parentPath.replace(
            `${buildersDir}/`, ""
        );

        const pathParts = pathWithoutBase.split("/");

        const spec = readFileSync(join(builder.parentPath, builder.name));

        if (typeof builders[pathParts[0]] === "undefined") {
            builders[pathParts[0]] = {};
        }

        builders[pathParts[0]][pathParts[1]] = {
            builderSpec: spec.toString(),
            name: pathParts[0],
            version: pathParts[1]
        };
    }
};
