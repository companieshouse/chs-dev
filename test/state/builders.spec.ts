import { expect, jest } from "@jest/globals";
import fs, { Dirent } from "fs";
import { dirname } from "path";
import BuilderDockerComposeSpec from "../../src/model/BuilderDockerComposeSpec";
import { clearBuilders, getBuilder } from "../../src/state/builders";
import { generateBuilderSpec } from "../utils/docker-compose-spec";
import yaml from "yaml";

type EntryType = "file" | "directory" | "blockdevice"
    | "characterdevice" | "symboliclink" | "fifo" | "socket";

const direntMock: (entryName: string, path: string, entryType: EntryType) => Dirent = (entryName, path, entryType) => {
    return {
        name: entryName,
        path,
        parentPath: path,
        isFile: () => entryType.toLowerCase() === "file",
        isDirectory: () => entryType.toLowerCase() === "directory",
        isBlockDevice: () => entryType.toLowerCase() === "blockdevice",
        isCharacterDevice: () => entryType.toLowerCase() === "characterdevice",
        isSymbolicLink: () => entryType.toLowerCase() === "symboliclink",
        isFIFO: () => entryType.toLowerCase() === "fifo",
        isSocket: () => entryType.toLowerCase() === "socket"
    };
};

describe("getBuilder", () => {
    const readdirSyncSpy = jest.spyOn(fs, "readdirSync");
    const existsSyncSpy = jest.spyOn(fs, "existsSync");
    const readFileSyncSpy = jest.spyOn(fs, "readFileSync");

    beforeEach(() => {
        jest.resetAllMocks();

        readdirSyncSpy.mockReturnValue([
            direntMock("java", "local/builders", "directory"),
            direntMock("v1", "local/builders/java", "directory"),
            direntMock("builder.docker-compose.yaml", "local/builders/java/v1", "file"),
            direntMock("Dockerfile", "local/builders/java/v1", "file"),
            direntMock("v2", "local/builders/java", "directory"),
            direntMock("builder.docker-compose.yaml", "local/builders/java/v2", "file"),
            direntMock("builder.Dockerfile", "local/builders/java/v2", "file"),
            direntMock("Dockerfile", "local/builders/java/v2", "file"),
            direntMock("Dockerfile", "local/builders/java", "symboliclink"),
            direntMock("node", "local/builders", "directory"),
            direntMock("v1", "local/builders/node", "directory"),
            direntMock("builder.docker-compose.yaml", "local/builders/node/v1", "file"),
            direntMock("Dockerfile", "local/builders/node/v1", "file"),
            direntMock("v2", "local/builders/node", "directory"),
            direntMock("builder.docker-compose.yaml", "local/builders/node/v2", "file"),
            direntMock("builder.Dockerfile", "local/builders/node/v2", "file"),
            direntMock("Dockerfile", "local/builders/node/v2", "file"),
            direntMock("Dockerfile", "local/builders/node", "symboliclink")
        ]);

        existsSyncSpy.mockReturnValue(true);

        readFileSyncSpy.mockImplementation((path, options) => {
            const builderPath = dirname(path as string);

            return Buffer.from(generateBuilderSpec(builderPath, builderPath.includes("v2")), "utf8");
        });
    });

    it("reads builders directory", () => {
        getBuilder("./", "java", undefined);

        expect(readdirSyncSpy).toHaveBeenCalledWith(
            "local/builders",
            {
                recursive: true,
                withFileTypes: true
            }
        );

        clearBuilders();
    });

    it("reads each file in", () => {
        getBuilder("./", "java", undefined);

        expect(readFileSyncSpy).toHaveBeenCalledTimes(4);
        expect(readFileSyncSpy).toHaveBeenCalledWith(
            "local/builders/java/v1/builder.docker-compose.yaml"
        );
        expect(readFileSyncSpy).toHaveBeenCalledWith(
            "local/builders/java/v2/builder.docker-compose.yaml"
        );
        expect(readFileSyncSpy).toHaveBeenCalledWith(
            "local/builders/node/v1/builder.docker-compose.yaml"
        );
        expect(readFileSyncSpy).toHaveBeenCalledWith(
            "local/builders/node/v2/builder.docker-compose.yaml"
        );
    });

    it("returns latest builder when version is undefined", () => {
        const result = getBuilder("./", "java", undefined) as BuilderDockerComposeSpec;

        expect(result.name).toBe("java");
        expect(result.version).toBe("v2");
        expect(result.builderSpec).toEqual(
            generateBuilderSpec("local/builders/java/v2", true)
        );
    });

    for (const version of ["v1", "v2"]) {
        it(`returns correct builder when version is ${version}`, () => {
            const result = getBuilder("./", "node", version) as BuilderDockerComposeSpec;

            expect(result.name).toBe("node");
            expect(result.version).toBe(version);
            expect(result.builderSpec).toEqual(
                generateBuilderSpec(`local/builders/node/${version}`, version === "v2")
            );
        });
    }

    it("calls readdirSync once when called multiple times - i.e. results are cached", () => {
        clearBuilders();

        getBuilder("./", "java", undefined);
        getBuilder("./", "java", undefined);

        expect(readdirSyncSpy).toHaveBeenCalledTimes(1);
    });

    it("returns repository builder", () => {
        const expectedBuilderSpec = yaml.stringify({
            services: {
                "<service>": {
                    pull_policy: "build",
                    build: {
                        context: "<absolute_repository_path>",
                        args: {
                            // eslint-disable-next-line no-template-curly-in-string
                            SSH_PRIVATE_KEY: "${SSH_PRIVATE_KEY}",
                            // eslint-disable-next-line no-template-curly-in-string
                            SSH_PRIVATE_KEY_PASSPHRASE: "${SSH_PRIVATE_KEY_PASSPHRASE}"
                        }
                    },
                    develop: {
                        watch: [
                            {
                                path: ".touch",
                                action: "rebuild"
                            }
                        ]
                    }
                }
            }
        });

        const result = getBuilder("./", "repository", undefined);

        expect(result).toEqual({
            name: "repository",
            version: "v1",
            builderSpec: expectedBuilderSpec
        });
    });

    it("throws error when builder version does not exist", () => {
        expect(() => getBuilder("./", "java", "v3")).toThrowErrorMatchingSnapshot();
    });

    it("throws error when builder does not exist", () => {
        expect(() => getBuilder("./", "javascript", undefined)).toThrowErrorMatchingSnapshot();
    });

    for (const testCase of [undefined, ""]) {

        it(`returns repository builder when builder name is "${testCase}"`, () => {
            const expectedBuilderSpec = yaml.stringify({
                services: {
                    "<service>": {
                        pull_policy: "build",
                        build: {
                            context: "<absolute_repository_path>",
                            args: {
                                // eslint-disable-next-line no-template-curly-in-string
                                SSH_PRIVATE_KEY: "${SSH_PRIVATE_KEY}",
                                // eslint-disable-next-line no-template-curly-in-string
                                SSH_PRIVATE_KEY_PASSPHRASE: "${SSH_PRIVATE_KEY_PASSPHRASE}"
                            }
                        },
                        develop: {
                            watch: [
                                {
                                    path: ".touch",
                                    action: "rebuild"
                                }
                            ]
                        }
                    }
                }
            });

            const result = getBuilder("./", testCase, undefined);

            expect(result).toEqual({
                name: "repository",
                version: "v1",
                builderSpec: expectedBuilderSpec
            });
        });
    }
});
