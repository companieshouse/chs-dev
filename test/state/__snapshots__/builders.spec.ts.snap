// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBuilder throws error when builder does not exist 1`] = `"No builder found for javascript builder"`;

exports[`getBuilder throws error when builder version does not exist 1`] = `"No version v3 of java builder"`;
